### PR TITLE
Linear LR warmup then cosine annealing across all 4 CFD benchmarks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Our current schedule starts cosine annealing from epoch 1. A brief linear warmup (e.g. 10 epochs ramping from 0 to peak LR) before entering cosine annealing prevents destructive early updates when model weights are randomly initialized, and is well-established in large-scale transformer training (e.g. GPT, ViT). Physics surrogate models with high-variance pressure targets may benefit especially — the early epochs set the scale of the internal representations, and noisy large steps can lock in a poor basin. A short warmup followed by the existing cosine schedule could yield both faster and deeper convergence.

**Cross-dataset scope:** TandemFoil · TandemFoil Paper · AirfRANS · DrivAerML

---

## Instructions to the Student

Add a `--warmup-epochs` integer flag (default `0`, meaning no warmup) to `train.py`. When `--warmup-epochs N` is set:

1. For the first N epochs, scale the learning rate linearly from `lr * 1e-3` (near zero) to the full `lr`.
2. After epoch N, switch to the existing cosine annealing schedule (CosineAnnealingLR with T_max as currently configured), starting from the full LR.

Implementation hint: use `torch.optim.lr_scheduler.LinearLR` for the warmup followed by `CosineAnnealingLR`, chained with `SequentialLR`. Alternatively, implement a single lambda scheduler that does:
```python
def lr_lambda(epoch):
    if epoch < warmup_epochs:
        return (epoch + 1) / warmup_epochs  # linear ramp
    else:
        # cosine from peak to 0 over remaining epochs
        progress = (epoch - warmup_epochs) / max(1, total_epochs - warmup_epochs)
        return 0.5 * (1 + math.cos(math.pi * progress))
```

Test with `--warmup-epochs 10` on all datasets. (If TandemFoil with T_max=10 and warmup=10 proves odd, try warmup=5 as an alternative for TF only.)

Use `--wandb_group linear-warmup-scheduler` on all runs.

---

### TandemFoil

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --warmup-epochs 10 \
  --wandb_group linear-warmup-scheduler
```

### TandemFoil Paper

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 150 --no-use-ema \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --warmup-epochs 10 \
  --wandb_group linear-warmup-scheduler
```

### AirfRANS

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --warmup-epochs 10 \
  --wandb_group linear-warmup-scheduler
```

### DrivAerML

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --warmup-epochs 10 \
  --wandb_group linear-warmup-scheduler
```

---

## Current Baselines (lower is better)

| Dataset | Metric | Current Best | PR |
|---|---|---|---|
| TandemFoil | `val_primary/surface_pressure_mae` | 22.537 | #2924 |
| TandemFoil Paper | `val_primary/field_mse` | 0.00434 | #2979 |
| AirfRANS | `val_primary/surface_mse` | 0.000482 | #2951 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | 3.997% | #2898 |

---

## Reporting

When done, post results as a PR comment with:
- W&B run IDs for all runs
- Best val metric per dataset (warmup vs no-warmup baseline)
- Whether warmup beat baseline on each dataset
- LR curves if possible (warmup ramp visible in W&B?)

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
